changes/2024-TBD_cache-across-hierarchy-keyrings/change.md

```diff
@@ -0,0 +1,613 @@
+However, an instance of the Hierarchy Keyring
+can only ever call KMS with one KMS Relationship,
+which is, at least partly,
+configured on the KMS Client determined
+at the Hierarchy Keyring's construction.
+
+By KMS Relationship, we mean any or all of the following:
+
+- KMS Configuration
+- Credentials used when creating the KMS Client, and thus
+  used when calling KMS
+- Other properties of the KMS Client, such as the region,
+  or request headers
+
+The Local Cryptographic Material Cache of
+the Hierarchy Keyring instance is then only
+populated with Branch Keys that correspond with
+that KMS relationship.
+
+Which is appropriate,
+as it is clear under what KMS relationship
+a Branch Key is accessed.
+
+However,
+the Hierarchy Keyring,
+and it's Key Store,
+have a runtime cost,
+exerting memory pressure
+and, without manual optimization,
+requiring at least 2 TLS handshakes
+when first serving a request
+(TLS to KMS & TLS to DDB).
+
+Additionally,
+the local Cryptographic Materials Cache
+exerts some runtime cost,
+particularly in a multi-threaded environment,
+when a background worker thread MAY be refreshing
+or pruning entries of the cache.
+
+For MPL Consumers that MUST work with Branch Keys
+under different KMS Relationships,
+this runtime cost adds up.
+
+These MPL Consumers MAY end up establishing
+a LRU Cache of Hierarchy Keyrings.
+Which, while workable, is sub-optimal,
+and clearly makes the Hierarchy Keyring,
+in these conditions,
+"Hard to Use".
+
+The objective, with these changes,
+is to make the Hierarchy Keyring
+"Easy to Use" in a multiple KMS Relationship
+environment.
+
+## Summary
+
+<!-- TODO: Remove SIM before merging to mainline -->
+
+Relevant [sim](https://sim.amazon.com/issues/CrypTool-5311).
+
+To improve the MPL Consumer data key caching experience,
+when using the Hierarchy Keyring,
+we need to allow caching across Key Stores/KMS Clients/KMS Keys.
+
+To facilitate Caching across Key Stores/KMS Clients/KMS Keys,
+we MUST break the Cryptographic Materials Cache (CMC)
+out of the Hierarchy Keyring.
+
+By allowing MPL Consumers to provide an already initialized CMC
+to the Hierarchy Keyring at construction,
+the CMC MAY cache Branch Keys protected by different
+KMS Relationships.
+
+This simplifies Multiple KMS Relationship MPL Consumers,
+as they do not need to stand up a LRU Cache of Hierarchy Keyrings.
+
+Instead, they may maintain one CMC.
+They still create a Hierarchy Keyring instance per KMS Relationship,
+and they MUST use the correct Keyring to retrieve material
+from the Cache.
+
+But they only need to maintain the common cache.
+
+In the future, the CachingCMM will be introduced to
+Crypto Tool's Dafny products.
+The CachingCMM and the Hierarchy Keyring both consume CMCs.
+Thus, it will be possible to provide a CMC to both a
+Hierarchy Keyring & a CachingCMM.
+This cache identifier design MUST consider
+how the cache entries of multiple Hierarchy Keyrings
+and CachingCMMs will be appropriately separated.
```

Comment:
If this spec change is going to propose a change to caching cmm cache id formulas we should say that this cahnge also impacts the caching cmm up above or say that it is out of scope in the below section.

Reply:
We are updating the spec to update the cachingCMM cache identifiers.

Currently, we're only focussing on making the changes to the MPL. Eventually, everything is going to be moved to Dafny anyway.

If a case can be made for other native libraries, we can incorporate that.

Reply:
Update: Native library cache identifiers can NEVER collide with the new proposed cache identifiers.